### PR TITLE
Zypper repository rewrite

### LINKED ISSUE #939: zypper_repository dont changes repo url
##### Summary:

The module zypper_repository dont changes repositories if the given URL has changed.
##### Steps To Reproduce:

First I added a repo with

```
- name: add repository
  zypper_repository: name='OurRepo' repo='file://path/to/repo' state=present
```

which worked as expected. On the second run nothing changes.

Then we switched  to a http-repo, so I just changed the repo-url:

```
- name: add repository 
  zypper_repository: name='OurRepo' repo='http://rpm.exampe.org/repo/' state=present
```

but ansible didnt changed something until I removed the repo by hand and run it again.

Maybe there is something wrong with [L150](https://github.com/ansible/ansible-modules-extras/blob/devel/packaging/os/zypper_repository.py#L150) (I guess this part was for comparing if the repo-url/-name has changed) , or maybe i missunderstand what the 'expected behavior' should be.
##### Expected Results:

TASK: changed
on first run after edit.
##### Actual Results:

TASK: ok

### LINKED ISSUE #1450: zypper_repository: add "priority" argument
Issue Type: Feature Idea

zypper can set a priority for each added repository. Packages will be taken from the repository with the highest priority (smallest number). See also https://en.opensuse.org/SDB:Zypper_manual_%28plain%29

This functionality should be available through ansible as well. Add an arguement `priority` to the zypper_repository module.

### LINKED ISSUE #1633: Zypper repository: change auto-refresh on existing repositories
##### Summary:

I wanted to update an pre-existing repository with `zypper_repository` to set the autorefresh feature to on/off (depending on the repository).
##### Steps To Reproduce:

All four repositories do currently exist after a fresh install, all of them has "refresh => no".

```
$> zypper repos --uri
# | Alias                 | Name                  | Enabled | GPG Check | Refresh | URI                                                              
--+-----------------------+-----------------------+---------+-----------+---------+------------------------------------------------------------------
1 | distro-non-oss        | openSUSE Non-OSS      | Yes     | (r ) Yes  | No      | http://download.opensuse.org/distribution/leap/42.1/repo/non-oss/
2 | distro-oss            | openSUSE OSS          | Yes     | (r ) Yes  | No      | http://download.opensuse.org/distribution/leap/42.1/repo/oss/    
3 | distro-update-non-oss | distro-update-non-oss | Yes     | ( p) Yes  | No      | http://download.opensuse.org/update/leap/42.1/non-oss/           
4 | distro-update-oss     | openSUSE OSS Update   | Yes     | (r ) Yes  | No      | http://download.opensuse.org/update/leap/42.1/oss/ 
```

Then I execute the following tasks.

```
- name: auto-refresh on update repositories
  zypper_repository: name={{ item.name }} repo={{ item.repo }} refresh=yes state=present
  with_items:
    - { name: distro-update-non-oss, repo: "http://download.opensuse.org/update/leap/42.1/non-oss/" }
    - { name: distro-update-oss, repo: "http://download.opensuse.org/update/leap/42.1/oss/" }

- name: auto-refresh off on non-update repositories
  zypper_repository: name={{ item.name }} repo={{ item.repo }} refresh=no state=present                                                                                        
  with_items:
    - { name: distro-non-oss, repo: "http://download.opensuse.org/distribution/leap/42.1/non-oss/" }
    - { name: distro-oss, repo: "http://download.opensuse.org/distribution/leap/42.1/oss/" }

```

I always get success (ok=2, changed=0)
##### Expected Results:

My expectation was that the first two repositories do now have "refresh => yes".
##### Actual Results:

The repositories keep on "refresh => no".

### PR DESCRIPTION
**Pull request by robinro**
##### ISSUE TYPE
- Bugfix Pull Request
##### COMPONENT NAME

packaging/os/zypper_repository.py
##### ANSIBLE VERSION

devel
##### SUMMARY

This continues the rewrite proposed in #1515. But expand it:
- add properties `enabled` and `priority`
- allow changing of one property and correctly report changed
- allow overwrite of multiple repositories by alias and URL
- cleanup of unused code and more structuring
- fixes #1515 (rewrite this is based on)
- fixes #939 (change repo url) 
- fixes #1633 (change auto-refresh)
- fixes #1450 (add prio)

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @robinro. To the current maintainers, @matze please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
Thanks @robinro for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by robinro** [comment]
ready_for_review

**Comment by gregdek** [comment]
Thanks @robinro. To the current maintainers, @matze please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by matze** [comment]
In principle I'd say shipit, however the removal of support for older zypper versions in 58f1021 should be communicated appropriately.

**Comment by gregdek** [comment]
Thanks again to @robinro for this PR, and thanks @matze for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by robinro** [comment]
Thanks @matze 

Do you know how old zypper must be to require the old setup?
According to
https://en.opensuse.org/SDB:Zypper_usage_11.3#XML_output
the xml output was there in 11.3 already. Then also SLES 11.2 should have probably had it.

**Comment by robinro** [comment]
According to https://en.opensuse.org/openSUSE:Zypper_versions you have to go to opensuse 11.0 to have the old zypper version (0.x). In SLES I unterstand that 11.\* is fine, see https://www.suse.com/LinuxPackages/packageRouter.jsp?product=server&version=11&architecture=i386&package_name=index_all

@paavoap: Your my go-to guy for old SLES versions :) This PR would remove support for zypper 0.x. Does one of your older SLES versions still use it? Can you confirm that SLES 11 is already on zypper 1.x?

I added a preliminary note to the requirements. It should be updated once we know the SLES version needed.

**Comment by matze** [comment]
> Do you know how old zypper must be to require the old setup?

That does not matter if it breaks at least one person's setup. As I said, I am not against that change itself but it should be communicated appropriately.

**Comment by robinro** [comment]
@matze is a note in the requirements enough to communicate it?

**Comment by matze** [comment]
I guess that's enough.

Robin Roth notifications@github.com schrieb am Do., 14. Apr. 2016, 08:06:

**Comment by paavoap** [comment]
@robinro Fresh install from DVD

```
# cat /etc/SuSE-release
SUSE Linux Enterprise Server 11 (x86_64)
VERSION = 11
PATCHLEVEL = 0
# zypper --version
zypper 1.0.8
```

**Comment by robinro** [comment]
@paavoap Thanks! So this aligns with my guess in bad0162

File: packaging/os/zypper_repository.py [hunk 1]
```diff
@@ -58,16 +58,30 @@
...
+        description:
+            - Overwrite multiple repository entries, if repositories with both name and
+              URL already exist.
+        required: false
+        default: "no"
+        choices: [ "yes", "no" ]
+        version_added: "2.1"
+requirements: 
+    - "zypper  >= 1.0"
+    - "openSuSE >= 11.1"
```
Comment (alxgu) [inline packaging/os/zypper_repository.py]: opensuse or sles are not really requirements. That should be hints.
- "zypper >= 1.0" # included in openSuSE >= 11.1 or SuSE Linux Enterprise Server/Desktop >= 11.0

Comment (robinro) [inline packaging/os/zypper_repository.py]: sounds reasonable, added it

**Comment by resmo** [comment]
I am not able the tests but based on @robinro comments on IRC this PR is tested and tests extended in https://github.com/ansible/ansible/pull/15387.

**Comment by robinro** [comment]
integration test output

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
Monday 18 April 2016  17:51:07 +0200 (0:00:00.095)       0:00:00.095 *********\* 
ok: [myhostname]

TASK [test_zypper_repository : Delete] *****************************************
Monday 18 April 2016  17:51:07 +0200 (0:00:00.533)       0:00:00.629 *********\* 
changed: [myhostname]

TASK [test_zypper_repository : Add repo] ***************************************
Monday 18 April 2016  17:51:08 +0200 (0:00:00.289)       0:00:00.918 *********\* 
changed: [myhostname]

TASK [test_zypper_repository : debug] ******************************************
Monday 18 April 2016  17:51:08 +0200 (0:00:00.300)       0:00:01.219 *********\* 
ok: [myhostname] => {
    "zypper_result": {
        "changed": true, 
        "repodata": {
            "alias": "test", 
            "autorefresh": "1", 
            "enabled": "1", 
            "gpgcheck": "1", 
            "name": null, 
            "priority": null, 
            "url": "http://dl.google.com/linux/chrome/rpm/stable/x86_64"
        }, 
        "state": "present"
    }
}

TASK [test_zypper_repository : verify repo addition] ***************************
Monday 18 April 2016  17:51:08 +0200 (0:00:00.082)       0:00:01.302 *********\* 
ok: [myhostname]

TASK [test_zypper_repository : Add repo again] *********************************
Monday 18 April 2016  17:51:08 +0200 (0:00:00.074)       0:00:01.377 *********\* 
ok: [myhostname]

TASK [test_zypper_repository : verify no change on second install] *************
Monday 18 April 2016  17:51:08 +0200 (0:00:00.292)       0:00:01.670 *********\* 
ok: [myhostname]

TASK [test_zypper_repository : Change repo URL] ********************************
Monday 18 April 2016  17:51:08 +0200 (0:00:00.072)       0:00:01.743 *********\* 
changed: [myhostname]

TASK [test_zypper_repository : Verify change on URL only change] ***************
Monday 18 April 2016  17:51:09 +0200 (0:00:00.423)       0:00:02.166 *********\* 
ok: [myhostname]

PLAY RECAP *********************************************************************
myhostname        : ok=9    changed=3    unreachable=0    failed=0

**Comment by resmo** [comment]
great @robinro thanks for the test results

**Pull request merged**